pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/BaseRoutingTableBuilder.java

```diff
@@ -39,15 +47,28 @@
   protected final Random _random = new Random();
   private BrokerMetrics _brokerMetrics;
   private String _tableName;
+  private boolean _enableDyanmicComputing;
 
   // Set variable as volatile so all threads can get the up-to-date routing tables
-  private volatile List<Map<String, List<String>>> _routingTables;
+  // Routing tables are used for storing pre-computed routing table
+  protected volatile List<Map<String, List<String>>> _routingTables;
+
+  // A mapping of segments to servers is used for dynamic routing table building process
+  protected volatile Map<String, List<String>> _segmentToServersMap;
 
   @Override
-  public void init(
-      Configuration configuration, TableConfig tableConfig, ZkHelixPropertyStore<ZNRecord> propertyStore, BrokerMetrics brokerMetrics) {
+  public void init(Configuration configuration, TableConfig tableConfig, ZkHelixPropertyStore<ZNRecord> propertyStore,
+      BrokerMetrics brokerMetrics) {
     _tableName = tableConfig.getTableName();
     _brokerMetrics = brokerMetrics;
+
+    // Enable dynamic routing when the config is explicitly set
+    RoutingConfig routingConfig = tableConfig.getRoutingConfig();
+    if (routingConfig != null) {
+      Map<String, String> routingOption = routingConfig.getRoutingTableBuilderOptions();
+      _enableDyanmicComputing = Boolean.valueOf(routingOption.get(RoutingConfig.ENABLE_DYNAMIC_COMPUTING_KEY));
+      LOGGER.info("Dynamic routing table computation is enabled for table {}", _tableName);
+    }
   }
 
   protected static String getServerWithLeastSegmentsAssigned(List<String> servers,
```

Comment:
Not related to this change, but can be optimized by changing it to **assignSegmentToServerWithLeastSegmentsAssigned**

Reply:
i will handle this on the separate pr